build: split hash checking from the download step

jszakmeister opened the pull request:
It turns out that `file(DOWNLOAD ...)` is not very user friendly with
it's error message, and only supports MD5 on v2.8.10 of CMake (the
default for Ubuntu 12.04).  If CMake is built without SSL support,
users are left hanging with a message that the hashes don't match.

It turns out that `file(SHA1 ...)` exists in v2.8.10, and we
can use that to compute the hash ourselves.  So this splits the hash
checking into a separate step, where we can provide some additional
advice if the SHA1 is the hash for an empty file.  Additionally, it also
allows us to drop the MD5 hashes and maintain only SHA1 hashes for our
dependencies.

jszakmeister committed:
build: split hash checking from the download step

It turns out that `file(DOWNLOAD ...)` is not very user friendly with
it's error message, and only supports MD5 on v2.8.10 of CMake (the
default for Ubuntu 12.04).  If CMake is built without SSL support,
users are left hanging with a message that the hashes don't match.

It turns out that `file(SHA1 ...)` exists in v2.8.10, and we
can use that to compute the hash ourselves.  So this splits the hash
checking into a separate step, where we can provide some additional
advice if the SHA1 is the hash for an empty file.  Additionally, it also
allows us to drop the MD5 hashes and maintain only SHA1 hashes for our
dependencies.

fwalch commented:
Very nice! LGTM

Minor idea: use some human-readable string (e.g. `SKIP`) instead of the `NULL_SHA1` to skip hash validation.

jszakmeister commented:
> Minor idea: use some human-readable string (e.g. SKIP) instead of the NULL_SHA1 to skip hash validation.

`NULL_SHA1` is actually the name of the variable.  It's a string of 0's for the SHA1.  I usually do `viwr0` on the sha1 itself, which turns it all to 0's.  But we could use something else too.

justinmk commented on third-party/cmake/DownloadAndExtractFile.cmake:
```diff
@@ -85,6 +75,25 @@ if(NOT status_code EQUAL 0)
+  message(FATAL_ERROR
+    "Failed to download ${URL}.  The file is empty and likely means CMake "
+    "was built without SSL support.  Please use a version of CMake with "
+    "proper SSL support.")
```
link to https://github.com/neovim/neovim/wiki/Building-Neovim#build-prerequisites ?

jszakmeister replied:
I don't see that link offering anything the message doesn't.  If there was more there that would help with this specific problem it would seem worthwhile, but there's not (other than a link to a ticket, which says the same thing essentially).

justinmk replied:
The comment link has steps to build cmake. Up to you though.

justinmk commented on third-party/cmake/DownloadAndExtractFile.cmake:
```diff
@@ -85,6 +75,25 @@ if(NOT status_code EQUAL 0)
+    "proper SSL support.")
+elseif((NOT EXPECTED_SHA1 STREQUAL NULL_SHA1) AND
+       (NOT EXPECTED_SHA1 STREQUAL ACTUAL_SHA1))
+  # Wasn't a NULL SHA1 and we didn't match, so we fail.
```
I tend to encourage omitted comments if there's a nearby user-facing message that is mostly equivalent.

jszakmeister replied:
The user facing message is for something else (the sha1 of an empty file).  This is about the all zeros sha1, which we use as a wildcard.

justinmk commented:
LGTM

jszakmeister committed:
build: allow SKIP or skip to be used as the SHA1

This will skip hash checking, just like the all zeros hash.

jszakmeister commented:
@fwalch Once I got what you were saying through my thick head, I added the ability to use "SKIP" or "skip" as the SHA1.

@justinmk I added the url.  The build related bits were kind of buried, so I may just pull out the relevant portion and put it next to the CMake stuff.

I'll merge once I see Travis pass.
